### PR TITLE
Pass owner to constructor.

### PR DESCRIPTION
**Pull request by rwjblue**
Ensure that services and other injected things work properly from the constructor.

### CODE REVIEW

File: tests/integration/components/sparkles-component-test.js [hunk 1]
```diff
@@ -221,5 +242,24 @@ module('Integration | Component | sparkles-component', function(hooks) {
...
+      }
+    });
+    this.owner.register(
+      'template:components/under-test',
+      hbs`<p>Environment: {{this.environment}}</p>`
+    );
+    await render(hbs`<UnderTest />`);
+    assert.dom('p').hasText('Environment: test');
+
+    assert.deprecationsInclude((`must call super with all arguments in the constructor for UnderTest`));
```
Comment (pzuraq) [inline tests/integration/components/sparkles-component-test.js]: double parens here needed?

**Review by pzuraq — APPROVED** [review]
👍 generally looks good. Only thing I would say is it seems weird for `owner` to come second, I imagine it'll probably be first in general with framework objects in the future, but I can also understand why you'd want to do this without making breaking changes.

File: tests/integration/components/sparkles-component-test.js (hunk 1 above)
Comment (rwjblue) [reply tests/integration/components/sparkles-component-test.js]: No its a mistake, good catch

**Comment by rwjblue** [comment]
> Only thing I would say is it seems weird for owner to come second, I imagine it'll probably be first in general with framework objects in the future,

Totally agree. My thought was to force it to be two args first, then swap them in a public API in a major version bump. Which _should_ be easier...

**Comment by buschtoens** [comment]
I just had an idea, which might be really stupid, but I would like your opinion on it. What if, instead of passing `owner` as an argument to the constructor, we use some global state?

Since `super` has to be called synchronously, we can be certain that nobody can interfere.

```ts
export const __OWNER_REFERENCE: { owner?: ApplicationInstance } = Object.create(null);

export default class SparklesComponentManager {
  // ..

  createComponent(Klass: typeof SparklesComponent, args: ComponentManagerArgs): CreateComponentResult {
    __OWNER_REFERENCE.owner = this.owner;
    let instance = new Klass(args.named);
	__OWNER_REFERENCE.owner = undefined;
    return instance as CreateComponentResult;
  }

  // ...
}
```

```js
import { setComponentManager } from '@ember/component';
import { setOwner } from '@ember/application';
import { __OWNER_REFERENCE } from 'sparkles-component/component-managers/sparkles';

class SparklesComponent<T = object> {
  constructor(public args: T) {
    if (__OWNER_REFERENCE.owner) {
      setOwner(this, __OWNER_REFERENCE.owner);
    }
  }

  // ...
}
```

**Comment by rwjblue** [comment]
Yep, definitely would work. I'm mostly trying to match the API of the upcoming glimmer component RFC, which _currently_ has two args (owner and args). I'll cross link the RFC once its officially submitted (hopefully in the next day or so).

**Comment by buschtoens** [comment]
Is there anything holding this back? Can I help?

**Comment by pzuraq** [comment]
We decided to focus on GlimmerComponent implementation, since this would either not match the API, or be a breaking change

File: addon/index.ts [hunk 2]
```diff
@@ -1,7 +1,12 @@
 import { setComponentManager } from '@ember/component';
+import { setOwner } from '@ember/application';
 
 class SparklesComponent<T = object> {
-  constructor(public args: T) {}
+  constructor(public args: T, owner: any | undefined) {
```
Comment (jordpo) [inline addon/index.ts]: This is swapped now I believe

Comment (mike-north) [reply addon/index.ts]: should be
```suggestion
  constructor(public args: T, owner?: {}) {
```
to describe the following constraints:
(a) `undefined` need not be passed explicitly under any circumstances
(b) if an argument is passed, it cannot be `null` or `undefined`

Comment (rwjblue) [reply addon/index.ts]: @mike-north - Seems good, should I use `object` or `{}`? (mostly just curious to the reasoning)

Comment (mike-north) [reply addon/index.ts]: @rwjblue 
```ts
let x: object = '';  // ❌
let x: {} = '';  // ✅
```
If you want to allow `string`, `number` `Symbol` and `boolean`, use `{}`
If you just want things that are assignable to JS `Objects`, use `object`

**Comment by rwjblue** [comment]
See https://github.com/rwjblue/sparkles-component/pull/33, closing in favor of using `@glimmer/component` directly.

**Pull request closed without merging**